File: UniquepathTopDown.java
class Solution {
    //tc=m*n
    //sc-m*n
    int[][] memo;
    public int uniquePaths(int m, int n) {
        if(m==0 || n==0) return 0;
        memo = new int [m][n];
        return helper(0, 0, m , n);
    }
    private int helper(int i,int j, int m,int n)
    {
        //basecase
        if(i == m-1 && j == n-1) return 1;
        if(i<0 || j<0 || i == m || j == n) return 0;
         //logic
         if(memo[i][j] == 0)
         {
         memo[i][j] = helper(i+1,j,m,n) + helper(i,j+1,m,n);
         }      
         return memo[i][j];

    }
}

File: unique_path_top_down.py
"""
Time Complexity : O(m*n) where m is the number of rows and n is the number of columns
Space Complexity : O(m*n) where m is the number of rows and n is the number of columns

Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No
"""


class Solution:
    def __init__(self):
        self.memo = None

    def uniquePaths(self, m: int, n: int) -> int:
        if m == 0 or n == 0:
            return 0
        self.memo = [[0 for _ in range(n)] for _ in range(m)]
        return self.helper(m, n, 0, 0)

    def helper(self, m, n, i, j):
        # base
        if i < 0 or i == m or j < 0 or j == n:
            return 0
        if i == m - 1 and j == n - 1:
            return 1

        # logic
        if self.memo[i][j] == 0:
            self.memo[i][j] = self.helper(
                m, n, i, j+1) + self.helper(m, n, i+1, j)
        return self.memo[i][j]
